docs: add coding agent guidelines (#9822)
buildbuddy-io/buildbuddy: AGENTS.md

> # BuildBuddy Coding Agent Guidelines

Reviewer: Most of this guide seems usable as general contribution guidelines (for humans as well). Maybe we'd want to title this "BuildBuddy contribution guidelines" and put it in `CONTRIBUTING.md`.

In terms of these various AGENTS files - historically we've avoided checking in any files that are specific to only some IDEs/workflows, such as `.vscode/` etc, so it seems like we should instead add all of these `AGENTS.md` etc. to `.gitignore` and then users can individually create symlinks to CONTRIBUTING.md if they want to use these tools.

Reviewer 2: +1 to generalizing this with CONTRIBUTING.md

## Build and Test

> - Use `./buildfix.sh` to format code

Reviewer: Maybe mention `./buildfix.sh -a` to fix Go dependencies.

- Use `bazel build` and `bazel test` to build and test the code
- To run a specific test in a file with Bazel, use `bazel test --test_filter=<function_name/subtest_name> //specific/package:target`

> - Use `--test_output=errors` to reduce the Bazel command verbosity

Reviewer: ```suggestion
- Use `--test_output=errors` to reduce test output verbosity
```

> - Use `--config=remote-minimal` to run the builds and tests remotely. Only compatible with Linux-specific targets

Reviewer: nit: the current wording suggests that only files named `*_linux_test.go` can run on `--config=remote-minimal`.

```suggestion
- Use `--config=remote-minimal` to run the builds and tests remotely (only compatible with targets that support Linux)
```

> - Use `genhtml` to explore Bazel coverage outputs

Reviewer: nit: maybe also add a bullet point about how to generate bazel coverage reports.

> - Use `bazel run //:gazelle` to update Bazel BUILD files

Reviewer: I think `buildfix.sh` takes care of this?

## Code Style

### Bazel

> - Never read or modify the `user.bazelrc` file.

Reviewer: If we renamed this file to CONTRIBUTING.md then this is the one point that I think would need to be removed, or maybe moved to a separate "Coding agents" section. More generally it might be better to describe what this is for, and let agents infer that they should not be touching this file.

- Prefer running builds and tests remotely, only fallback to local execution if there is an authentication issue.

### Proto

> - Keep all proto files in the `proto/` directory.

Reviewer: We have some protos not in this dir.

```suggestion
- Generally, most proto files are located in the `proto/` directory.
```

> - Vendor 3rd party proto files and update their imports as well as go_package option accordingly.

Reviewer: I don't think this generally applies. There are a lot of protos we don't vendor.

> - `proto/BUILD` file should be updated manually as Gazelle was configured to ignore it. Maintain the sorting order in the BUILD file.

Reviewer: The BUILD file isn't really sorted, currently.

```suggestion
- `proto/BUILD` file should be updated manually as Gazelle was configured to ignore it.
```

### Golang
- Avoid creating multiple files for a single package unless necessary. Acceptable cases are OS/Arch specific code
- Create an alias for proto package imports. For example `import inpb "github.com/buildbuddy-io/buildbuddy/proto/invocation"`
- Prefer table-driven tests with descriptive names
- To add new dependencies, first use `go get` to update `go.mod` and `go.sum`, then run `./tools/fix_go_deps.sh` to update `./deps.bzl`.

### React/Typescript
- New dependencies should be added using `pnpm`.

> - Run dev server using `bazel run server`. The relevant config flags can be set in `./config/buildbuddy.local.yaml`.

Reviewer: I would avoid mentioning the configuration for now until we clean it up a bit. Some of the config flags are set as `args` on the bazel target, and some are in the config file.

```suggestion
- To run a local development server for the OSS version of BuildBuddy, use `bazel run server`.
```

> - Run enterprise-version dev server using `bazel run enterprise/server`. The relevant config flags can be set in `./enterprise/config/buildbuddy.local.yaml`. Enterprise server depends on a Redis-server instance to run. Redis could be started with `docker run -d -p6379:6379 redis` or `redis-server` if installed locally.

Reviewer: No need to hyphenate "enterprise-version".

```suggestion
- Run a BuildBuddy enterprise development server using `bazel run enterprise/server`. Redis can be started with `docker run -d -p6379:6379 redis` or `redis-server` if installed locally. On Ubuntu, just installing redis should be sufficient (systemd should start the server automatically).
```

> - Additional flags could be given to the server instance via commandline arguments. For example, `bazel run enterprise/server -- --cache.detailed_stats_enabled=true` will enable detailed cache stats UI in the dev server.

Reviewer: ```suggestion
- Additional flags can be given to the server instance via command line arguments. For example, `bazel run enterprise/server -- --cache.detailed_stats_enabled=true` will enable the detailed cache stats UI in the dev server.
```
